Pebble: enable background javascript engine

danielegobbetti opened the pull request:
The background javascript engine runs in the same webview that runs the configuration page.
The engine can be toggled from the preferences, but only if experimental features are enabled.
No known regressions are present if background javascript is disabled.

Fixes #773 #467 #482 #327

cpfeiffer commented on app/src/main/java/nodomain/freeyourgadget/gadgetbridge/util/WebViewSingleton.java:
```diff
@@ -0,0 +1,275 @@
+                return json.getJSONObject("appKeys");
+            }
+        } catch (IOException | JSONException e) {
+            LOG.warn(e.getMessage());
```
Log the exception :)

cpfeiffer commented on app/src/main/java/nodomain/freeyourgadget/gadgetbridge/util/WebViewSingleton.java:
```diff
@@ -0,0 +1,275 @@
+            jsAppMessage.put("payload", outgoing);
+
+        } catch (Exception e) {
+            LOG.warn(e.getMessage());
```
You know what I mean :)

cpfeiffer commented on app/src/main/java/ru/gelin/android/weather/notification/ParcelableWeather2.java:
```diff
@@ -67,22 +79,105 @@ private ParcelableWeather2(Parcel in) {
+                    reconstructedForecast.put("list", list);
+
+                } catch (JSONException e) {
+                    e.printStackTrace();
```
I guess if we get any of the exceptions here, we should abort the whole thing, since it's unlikely it would work correctly, right?

danielegobbetti replied:
Not really, the weather is also used in other forms (not JSON-based)

cpfeiffer commented on app/src/main/java/nodomain/freeyourgadget/gadgetbridge/util/WebViewSingleton.java:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    public static void disposeWebView() {
+        if (internetHelperBound) {
```
Does re-setting up the whole webview work after it has been disposed once?

danielegobbetti replied:
It works without issues on my device after multiple disconnects.

danielegobbetti commented on app/src/main/java/nodomain/freeyourgadget/gadgetbridge/util/WebViewSingleton.java:
```diff
@@ -0,0 +1,275 @@
+    public static CountDownLatch latch;
+    public static WebResourceResponse internetResponse;
+    public final static Messenger internetHelperListener = new Messenger(new IncomingHandler());
+
```
I could not reduce the amount of static fields without causing malfunction, unfortunately. Perhaps after merging to master someone could help.

cpfeiffer replied:
Alright.

danielegobbetti commented on app/src/main/java/nodomain/freeyourgadget/gadgetbridge/devices/EventHandler.java:
```diff
@@ -63,7 +63,7 @@
     void onAppDelete(UUID uuid);
 
-    void onAppConfiguration(UUID appUuid, String config);
+    void onAppConfiguration(UUID appUuid, String config, Integer id);
```
This change was made by @ashimokawa, so I'll let him answer here :)

danielegobbetti commented on app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/ExternalPebbleJSActivity.java:
```diff
@@ -17,376 +17,158 @@
+                    WebViewSingleton.runJavascriptInterface(currentDevice, currentUUID);
+                } else if (extras.getBoolean(START_BG_WEBVIEW, false)) {
+                    WebViewSingleton.ensureCreated(this);
+                    finish();
```
return was added but github did not pick it up and didn't hide this comment.

danielegobbetti commented:
I believe I addressed the comments that were addressable :)
Thanks for the review! Do not hesitate to dig deeper before merging, but I'm using this as daily driver since a long time without issues.
